STACpopulator/populator_base.py

```diff
@@ -84,21 +86,33 @@ def ingest(self) -> None:
         if not stac_collection_exists(self.stac_host, self.collection_id):
             LOGGER.info(f"Creating collection '{self.collection_name}'")
             pystac_collection = create_stac_collection(self.collection_id, self._collection_info)
-            post_collection(self.stac_host, pystac_collection)
+            post_stac_collection(self.stac_host, pystac_collection)
             LOGGER.info("Collection successfully created")
         else:
             LOGGER.info(f"Collection '{self.collection_name}' already exists")
-        # for item in self.crawler(self.catalog, **self._crawler_args):
-        #     stac_item = self.process_STAC_item(item)
-        #     self.post_item(stac_item)
 
-    def post_item(self, data: dict[str, dict]) -> None:
+        # Item ingestion loop
+        for item_name, item_data in self._ingest_pipeline:
+            LOGGER.info(f"Creating STAC representation for {item_name}")
+            stac_item = self.create_stac_item(item_name, item_data)
+            if self.validate_stac_item_cv(stac_item):
+                if post_stac_item(self.stac_host, self.collection_id, stac_item):
+                    LOGGER.info(f"{item_name} successfully posted")
+                else:
+                    LOGGER.error(f"Posting {item_name} failed")
+                    self.handle_ingestion_error("Posting Error", item_name, item_data)
+            else:
+                LOGGER.error(f"Validation failed for item {item_name}")
+                self.handle_ingestion_error("Validation Error", item_name, item_data)
+
+    @abstractmethod
+    def handle_ingestion_error(self, error: str, item_name: str, item_data: MutableMapping[str, Any]):
         pass
 
     @abstractmethod
-    def process_stac_item(self):  # noqa N802
+    def create_stac_item(self, item_name: str, item_data: MutableMapping[str, Any]) -> MutableMapping[str, Any]:
```

Comment:
Good question. Right now the `create_stac_collection` is in `stac_utils` because it does not contain any logic that might need to be overridden by any derived classes of `STACpopulatorBase`. But, that is not to say that in future there might not be a need to override some parts of the logic. Do you think we should move all the create/post stac item/collection functions to `STACpopulatorBase`?

Reply:
Not necessarily, I think there should be a different file for all the code where `requests` are sent to the stac host. In the future `stac_utils` might become too big and have too many different responsabilies which would hinder the code readability. I would just keep `url_validate` inside `stac_utils`. Also, the `create_stac_collection` could be in `STACpopulatorBase` as a non-abstract method and be changed later on if there's a need for that. Let me know what you think.